### ISSUE TITLE
Sniff unit tests fail because of the <exclude-pattern> behaviour

### ISSUE THREAD

**Issue by PatrickRose** [squizlabs/PHP_CodeSniffer#1169]
I have a sniff set up like so in my standard's ruleset.xml

```xml
<rule ref="StandardName.PHP.SniffName">
	<!--Have to use a negative lookahead as PHPCS doesn't have support for include patterns-->
	<exclude-pattern>path/that/includes/phpcs/directory</exclude-pattern>
</rule>
```

When I run the Sniff unit tests, the `.inc` files for testing are being excluded so my unit tests fail.

**Comment by aik099** [comment]
Which unit test suite are you using? As per my knowledge the built-in unit test suite doesn't support running tests for sniffs created in non-built-in standards.

**Comment by gsherwood** [comment]
I figured out what was happening here and have committed a change to the unit testing system that should fix this. I couldn't replicate the problem on the 3.0 branch because it does testing in a different way, but I put the change in there anyway to give it a bit of a speed bump.

Basically, PHPCS tests one sniff at a time, but processes the ruleset first and then restricts the list of returned sniffs to a single one. It's basically the same as using the `--sniffs` command line argument. Because it processed the ruleset, things like exclude patterns get set.

Now, it doesn't process the ruleset at all. It fakes the list of sniffs and uses all the normal defaults so that only the unit testing system is in charge of setting them.

Let me know if this doesn't end up working for you, but I was able to replicate your problem and this change does fix that specific exclude-pattern case. Thanks for reporting it.

**Comment by PatrickRose** [comment]
I'll give this a shot later today hopefully

@aik099 We aren't loading PHPCS via phar or composer so I just placed our sniffs in the relevant directory.

**Comment by PatrickRose** [comment]
Finally got around to this - works like a charm. Cheers!

**Comment by aik099** [comment]
A nasty side effect of this fix is that unit test runner now only looks at sniffs from standards listed in `installed_paths` config setting hence #1205 problem.

Somehow the `installed_paths` setting by default points to standards folder of PHP_CodeSniffer itself and that's why sniff discovery isn't broken in there. All custom test suites are out of luck however.
